teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,478 @@
+      * uri: resource uri of the git commit
+      * digest: revision digest
+      * alg: digest algorithm: sha256 etc.
+      * src_type: source VCS type: git, hg, etc.
+    properties:
+    src_type:
+      type: string
+    uri:
+      type: string
+    alg:
+      type: string
+    digest:
+      type: string
+```
+The required formats for src_type is one of `git`, `hg`, `svc`, `baz`, `cvs`, `perforce`. The source types can be easily extended when more VCS are supported. The required formats for uri should follow [ResourceURI](https://github.com/in-toto/attestation/blob/main/spec/field_types.md#field-type-definitions). The required format for alg is one of pre-defined algorithm names in intoto attestation [field type definitions](https://github.com/in-toto/attestation/blob/main/spec/field_types.md#field-type-definitions). The digest is the revision digest.
+
+#### OCI Images
+
+OCI image sources will have the following schema:
+```
+...
+  - name: CHAINS-OCI_IMAGE
+    type: object
+    description: |
+      The source distribution
+      * name: name of the built image
+      * digest: image digest
+      * alg: digest algorithm: sha256 etc.
+    properties:
+    name:
+      type: string
+    digest:
+      type: string
+    alg:
+      type: string
+```
+The `name` is the OCI image name, for example, `gcr.io/test/foo`. The `digest` will be the image digest in the form of `69e70a79f2d41ab5d637de98c1e0b055206ba40a8145e7bddb55ccc04e13cf8f`. The `alg` follows the same requirements for VCS `alg` field.
+
+#### Python
+
+TODO
+
+#### Maven
+
+TODO
+
+#### Go
+
+TODO
+
+#### NodeJS
+
+TODO
+
+#### Parsing Results and Params from TaskRuns
+
+The defined results and params are in object types, and will be retrieved from TaskRun params and results. From the syntax of how the results / params look, the type of those, whether they are strings, objects or arrays can be determined. We need to validate the targets’ structures against defined result and param structures. Taking VCS as an example, when a `CHAINS-VCS` is found, the target needs to be an object, and it has to have at least the fields in defined `CHAINS-VCS` schema: `src_type`, `uri`, `alg` and `digest`. If additional fields are in the target, they will be ignored. The field validation rule follows the same pattern of duck typing mentioned in [TEP-0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md). These individual fields will also be validated against supported formats.
+
+#### Storage of Signable Target Provenance Schema
+
+Each target provenance schema will be stored in a file as a serialized json object. They will be loaded and mounted into the Chains controller image at volumeMountPath, for example, `tekton/target_schemas`. A singleton variable will load them into structs for later use. The naming will be the type hinting marker. For example, VCS schema file will be called `chains-vcs` and it should be stored in the file in such format:
```

Comment:
> The naming will be the type hinting 

I'd recommend against this, since if we're using results these will be user controlled values.

Possibly a crazy idea (not even 100% sold on this myself), but perhaps provenance information shouldn't belong in results and instead should be their own independent field? 🤔

Reply:
Updated the proposal section to reflect what we discussed